chore: Re-simplify `IndexVersionType` and `MessageLevel` with c++20

Ryex opened the pull request:
Since we have c++20 we can keep the nice api surface without needing all the boilerplate overloads.

This sets our compiler requirements to gcc 11 and clang 13.
If we forgo the use of `using enum` we can drop to gcc 10 and clang 10 but that means using `MessageLevel::Enum::Unknown` for direct enum access

<!--
Hey there! Thanks for your contribution.

Please make sure that your commits are signed off first.
If you don't know how that works, check out our contribution guidelines: https://github.com/PrismLauncher/PrismLauncher/blob/develop/CONTRIBUTING.md#signing-your-work
If you already created your commits, you can run `git rebase --signoff develop` to retroactively sign-off all your commits and `git push --force` to override what you have pushed already.

Note that signing and signing-off are two different things!
-->

TheKodeToad commented on launcher/MessageLevel.h:
```diff
@@ -2,26 +2,40 @@
+    using enum Enum;
+    constexpr MessageLevel(Enum e = Unknown) : m_type(e) {}
+    static MessageLevel fromString(const QString& type);
+    static MessageLevel fromQtMsgType(const QtMsgType& type);
```
```suggestion
    static MessageLevel fromName(const QString& type);
    static MessageLevel fromQt(const QtMsgType& type);
```

Ryex replied:
I'm willing to change the `fromString` but not the `fromQtMsgType`. `fromQt` isn't descriptive at all.

TheKodeToad replied:
That's why i chose it but somebody didn't like it iirc

Trial97 commented on launcher/modplatform/ModIndex.h:
```diff
@@ -107,7 +121,8 @@ struct IndexedVersion {
     {
-        auto release_type = version_type != IndexedVersionType::Unknown ? QString(" [%1]").arg(indexedVersionTypeToString(version_type)) : "";
+        auto release_type =
+            version_type != IndexedVersionType::Unknown ? QString(" [%1]").arg(version_type.toString()) : "";
```
```suggestion
            version_type.isValid() ? QString(" [%1]").arg(version_type.toString()) : "";
```

Trial97 reviewed (commented):
I still prefer the C-style approach for enums, but as long as we are consistent, I do not care so much.
So, regarding consistency, we should check other enums that we have to apply the same changes to them if needed(not necessary in this PR)

TheKodeToad commented on launcher/MessageLevel.h:
Why were this kept the same

Ryex replied:
because they modify the line thus and don't fit as members

TheKodeToad replied:
I'd go with:
```
MessageLevel::takeFromLine
MessageLevel::takeFromLauncherLine
```
(though the latter one is weird as a temp variable is always used to avoid the mutation)

TheKodeToad commented:
Also I just realised fromName was only changed in one place... and it doesn't have symmetry with toString... so maybe not a good idea (well I'd probably rename toString to getName but idk that's such a nitpick)

Ryex commented:
> Also I just realised fromName was only changed in one place... and it doesn't have symmetry with toString... so maybe not a good idea (well I'd probably rename toString to getName but idk that's such a nitpick)

push the change yourself if you want. I'm going to bed soon.

TheKodeToad commented:
It doesn't look like it compiles right now + conflicts with develop

Ryex commented:
ugg. that was a *mess*. not sure how I screwed that up so bad. only had to force push 5 time to fix the rebase...

Ryex committed:
refactor(MessageLevel): move and rename 'FromLine*' functions

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
chore: Re-simplify `IndexVersionType` and `MessageLevel` with c++20

This sets our compiler requirements to gcc 11 and clang 13. If we
forgo the use of `using enum` we can drop to gcc 10 and clang 10 but
that means using `MessageLevel::Enum::Unknown` for direct enum access

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>
